// skillspace-backend/controllers/booking.controller.js
import Booking from "../models/Booking.model.js";
import Review from '../models/Review.model.js';
import { google } from 'googleapis';

// 🟢 GOOGLE CALENDAR ENGINE PIPELINE
export const createGoogleCalendarEvent = async (booking, teacherTokens) => {
  const oauth2Client = new google.auth.OAuth2(
    process.env.GOOGLE_CLIENT_ID,
    process.env.GOOGLE_CLIENT_SECRET,
    process.env.GOOGLE_REDIRECT_URI
  );

  // Authenticate using the teacher's saved refresh token credentials
  oauth2Client.setCredentials({ refresh_token: teacherTokens });
  const calendar = google.calendar({ version: 'v3', auth: oauth2Client });

  const eventPayload = {
    summary: `SkillSwap: ${booking.skillTitle}`,
    description: `Your live peer-to-peer learning session is confirmed! Join your custom classroom here: http://localhost:3000/room/${booking._id}`,
    start: {
      dateTime: new Date(booking.sessionDate).toISOString(),
      timeZone: 'Asia/Kolkata',
    },
    end: {
      dateTime: new Date(new Date(booking.sessionDate).getTime() + 60 * 60 * 1000).toISOString(), // Default to 1 hour later
      timeZone: 'Asia/Kolkata',
    },
    attendees: [
      { email: booking.teacher.email },
      { email: booking.learner.email },
    ],
    reminders: {
      useDefault: false,
      overrides: [
        { method: 'popup', minutes: 15 },
        { method: 'email', minutes: 30 },
      ],
    },
  };

  // Insert event and automatically send invitation emails to both attendees
  await calendar.events.insert({
    calendarId: 'primary',
    resource: eventPayload,
    sendUpdates: 'all', 
  });
};

// POST /api/bookings
export const createBooking = async (req, res) => {
  try {
    const { swapRequest, teacher, skillTitle, sessionDate, timeSlot } = req.body;
    const learner = req.user._id;

    const slotTaken = await Booking.findOne({ teacher, sessionDate, timeSlot, status: 'confirmed' });
    if (slotTaken) {
      return res.status(400).json({ message: 'This time slot has already been booked' });
    }

    const booking = await Booking.create({
      swapRequest,
      teacher,
      learner,
      skillTitle,
      sessionDate,
      timeSlot,
    });
    res.status(201).json(booking);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

// PUT /api/bookings/:id
export const updateBookingStatus = async (req, res) => {
  try {
    const { status } = req.body;
    const booking = await Booking.findById(req.params.id);
    if (!booking) {
      return res.status(404).json({ message: 'Booking session not found' });
    }

    // Only the involved teacher or learner can modify
    const isTeacher = booking.teacher.toString() === req.user._id.toString();
    const isLearner = booking.learner.toString() === req.user._id.toString();

    if (!isTeacher && !isLearner) {
      return res.status(403).json({ message: 'Not authorized to change this booking' });
    } 

    booking.status = status;
    await booking.save();

    // ==========================================
    // 🟢 INTEGRATED STEP 5: AUTOMATED INVITATION TRIGGER
    // ==========================================
    if (status === 'confirmed') {
      // Re-query and populate full user objects to safely harvest emails and tokens
      const fullBooking = await Booking.findById(booking._id).populate('teacher learner');
      
      if (fullBooking.teacher.googleTokens?.refreshToken) {
        try {
          await createGoogleCalendarEvent(
            fullBooking, 
            fullBooking.teacher.googleTokens.refreshToken
          );
        } catch (apiErr) {
          // Log calendar errors gracefully without crashing the whole HTTP response thread
          console.error("Google Calendar failed to inject event invitation:", apiErr.message);
        }
      }
    }
    // ==========================================

    res.json(booking);

  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

// GET /api/bookings/mine
export const getMyBookings = async (req, res) => {
  try {
    const bookings = await Booking.find({
      $or: [{ teacher: req.user._id }, { learner: req.user._id }],
    })
      .populate('teacher', 'name email')
      .populate('learner', 'name email')
      .sort({ sessionDate: 1 });

    // Dynamically check if the user has already left a review for each booking
    const bookingsWithReviewStatus = await Promise.all(
      bookings.map(async (booking) => {
        const bookingObj = booking.toObject();
        
        if (booking.status === 'completed') {
          const reviewExists = await Review.findOne({
            booking: booking._id,
            reviewer: req.user._id,
          });
          bookingObj.hasReview = !!reviewExists;
        } else {
          bookingObj.hasReview = false;
        }
        
        return bookingObj;
      })
    );

    res.json(bookingsWithReviewStatus);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};
